jib-core/src/main/java/com/google/cloud/tools/jib/frontend/CredentialRetrieverFactory.java

```diff
@@ -0,0 +1,185 @@
+    return this;
+  }
+
+  /**
+   * Creates a new {@link CredentialRetriever} for retrieving credentials via a Docker credential
+   * helper, such as {@code docker-credential-gcr}.
+   *
+   * @param credentialHelperSuffix the credential helper executable suffix, following {@code
+   *     docker-credential-} (ie. {@code gcr} for {@code docker-credential-gcr})
+   * @return a new {@link CredentialRetriever}
+   */
+  public CredentialRetriever dockerCredentialHelper(String credentialHelperSuffix) {
+    return dockerCredentialHelper(
+        Paths.get(DockerCredentialHelperFactory.CREDENTIAL_HELPER_PREFIX + credentialHelperSuffix),
+        new DockerCredentialHelperFactory());
+  }
+
+  /**
+   * Creates a new {@link CredentialRetriever} for retrieving credentials via a Docker credential
+   * helper, such as {@code docker-credential-gcr}.
+   *
+   * @param credentialHelper the credential helper executable
+   * @return a new {@link CredentialRetriever}
+   * @see <a
+   *     href="https://github.com/docker/docker-credential-helpers#development">https://github.com/docker/docker-credential-helpers#development</a>
+   */
+  public CredentialRetriever dockerCredentialHelper(Path credentialHelper) {
+    return dockerCredentialHelper(credentialHelper, new DockerCredentialHelperFactory());
+  }
+
+  /**
+   * Creates a new {@link CredentialRetriever} that tries common Docker credential helpers to
+   * retrieve credentials based on the registry of the image, such as {@code docker-credential-gcr}
+   * for images with the registry as {@code gcr.io}.
+   *
+   * @return a new {@link CredentialRetriever}
+   */
+  public CredentialRetriever inferCredentialHelper() {
+    return inferCredentialHelper(new DockerCredentialHelperFactory());
+  }
+
+  @VisibleForTesting
+  CredentialRetriever inferCredentialHelper(
+      DockerCredentialHelperFactory dockerCredentialHelperFactory) {
+    String registry = imageReference.getRegistry();
+
+    List<String> inferredCredentialHelperSuffixes = new ArrayList<>();
+    for (String registrySuffix : COMMON_CREDENTIAL_HELPERS.keySet()) {
+      if (!registry.endsWith(registrySuffix)) {
+        continue;
+      }
+      String inferredCredentialHelperSuffix = COMMON_CREDENTIAL_HELPERS.get(registrySuffix);
+      if (inferredCredentialHelperSuffix == null) {
+        throw new IllegalStateException("No COMMON_CREDENTIAL_HELPERS should be null");
+      }
+      inferredCredentialHelperSuffixes.add(inferredCredentialHelperSuffix);
+    }
+
+    return () -> {
+      for (String inferredCredentialHelperSuffix : inferredCredentialHelperSuffixes) {
+        try {
+          return retrieveFromDockerCredentialHelper(
+              registry,
+              Paths.get(
+                  DockerCredentialHelperFactory.CREDENTIAL_HELPER_PREFIX
+                      + inferredCredentialHelperSuffix),
+              dockerCredentialHelperFactory);
+
+        } catch (NonexistentDockerCredentialHelperException ex) {
```

Comment:
I think you should also catch `NonexistentServerUrlDockerCredentialHelperException`, so that you try the next helper in the inferred list if you can't get the credentials from the first one in the list.

Reply:
I think we can leave that to another PR since it changes the behavior from the original `RetrieveRegistryCredentialsStep`. Filed #827